Libplanet/Blockchain/BlockChain.Swap.cs

```diff
@@ -0,0 +1,461 @@
+                        "different genesis.";
+                    throw new InvalidGenesisBlockException(Genesis.Hash, other.Genesis.Hash, msg);
+                }
+
+                _logger.Debug(
+                    "The branchpoint is #{BranchpointIndex} {BranchpointHash}.",
+                    branchpoint.Index,
+                    branchpoint
+                );
+
+                Block<T> oldTip = Tip ?? Genesis, newTip = other.Tip ?? other.Genesis;
+
+                ImmutableList<Block<T>> rewindPath =
+                    GetRewindPath(this, branchpoint.Hash);
+                ImmutableList<Block<T>> fastForwardPath =
+                    GetFastForwardPath(other, branchpoint.Hash);
+
+                // If there is no rewind, it is not considered as a reorg.
+                bool reorg = rewindPath.Count > 0;
+
+                _rwlock.EnterWriteLock();
+                try
+                {
+                    IEnumerable<Transaction<T>>
+                        GetTxsWithRange(BlockChain<T> chain, Block<T> start, Block<T> end)
+                        => Enumerable
+                            .Range((int)start.Index + 1, (int)(end.Index - start.Index))
+                            .SelectMany(x => chain[x].Transactions);
+
+                    // It assumes reorg is small size. If it was big, this may be heavy task.
+                    ImmutableHashSet<Transaction<T>> unstagedTxs =
+                        GetTxsWithRange(this, branchpoint, Tip).ToImmutableHashSet();
+                    ImmutableHashSet<Transaction<T>> stageTxs =
+                        GetTxsWithRange(other, branchpoint, other.Tip).ToImmutableHashSet();
+                    ImmutableHashSet<Transaction<T>> restageTxs = unstagedTxs.Except(stageTxs);
+                    foreach (Transaction<T> restageTx in restageTxs)
+                    {
+                        StagePolicy.Stage(this, restageTx);
+                    }
+
+                    Guid obsoleteId = Id;
+                    Id = other.Id;
+                    Store.SetCanonicalChainId(Id);
+                    _blocks = new BlockSet<T>(Policy.GetHashAlgorithm, Store);
+                    TipChanged?.Invoke(this, (oldTip, newTip));
+
+                    Store.DeleteChainId(obsoleteId);
+                }
+                finally
+                {
+                    _rwlock.ExitWriteLock();
+                }
+
+                renderSwap = () => RenderSwap(
+                    render: render,
+                    oldTip: oldTip,
+                    newTip: newTip,
+                    branchpoint: branchpoint,
+                    rewindPath: rewindPath,
+                    fastForwardPath: fastForwardPath,
+                    stateCompleters: completers);
+            }
+            finally
+            {
+                _rwlock.ExitUpgradeableReadLock();
+            }
+
+            return renderSwap;
```

Comment:
The `Swap()` method is seemingly used in a form like `Swarm(...)()`, which means `renderSwap`, the return value, is always immediately invoked once and only once.  Then, how about invoke it here before returning the function, and make `Swap()`'s return type `void`?  The current way has more opportunities to make a mistake because:

- The returned action could be never invoked by mistake.
- The returned action could be invoked twice or more. (E.g., `var swap = Swap(...); swap(); swap();`)

```suggestion
            renderSwap();
```

Reply:
IMO, separating `renderSwap()` call from `.Swap()` is the key of this PR because we want to broadcast block before 
 `renderSwap()`.

Reply:
I misread the code.  I overlooked the usage in the `Swarm<T>.ProcessFillBlocks()` method.  Fair enough.

Reply:
It would be better if we test whether the returned action does the rendering, e.g., check if nothing is rendered when the returned action is not invoked.